src/supervised/distance.jl

```diff
@@ -108,32 +108,154 @@ islipschitzcont_deriv(::L2DistLoss) = true
+    k::T   # k = 2π/circumference
+    function PeriodicLoss(circ::T)
+        circ > 0 || error("circumference should be strictly positive")
+        new(convert(T, 2π/circ))
+    end
+end
+function PeriodicLoss{T<:Number}(ε::T=1.0)
+    if T <: AbstractFloat
+        PeriodicLoss{T}(ε)
+    else # cast to Float64
+        PeriodicLoss{Float64}(Float64(ε))
+    end
+end
+
+value{T<:Number}(loss::PeriodicLoss, difference::T) = 1 - cos(difference*loss.k)
+deriv{T<:Number}(loss::PeriodicLoss, difference::T) = loss.k * sin(difference*loss.k)
+deriv2{T<:Number}(loss::PeriodicLoss, difference::T) = abs2(loss.k) * cos(difference*loss.k)
+function value_deriv{T<:Number}(loss::PeriodicLoss, difference::T)
+    dk = difference*loss.k
+    return 1-cos(dk), loss.k*sin(dk)
+end
+
+isdifferentiable(::PeriodicLoss) = true
+isdifferentiable(::PeriodicLoss, at) = true
+istwicedifferentiable(::PeriodicLoss) = true
+istwicedifferentiable(::PeriodicLoss, at) = true
+islipschitzcont(::PeriodicLoss) = true
+islipschitzcont_deriv(::PeriodicLoss) = true
+isconvex(::PeriodicLoss) = false
+isstronglyconvex(::PeriodicLoss) = false
+
+
+# ===========================================================
+# L(y, t) =  1/2 * (y-t)²    , if |y-t| < d
+# L(y, t) =  d⋅(|y-t| - d/2) , otherwise
+
+"""
+`HuberLoss <: DistanceLoss`
+
+Loss function commonly used for robustness to outliers
+"""
+type HuberLoss{T<:AbstractFloat} <: DistanceLoss
+    d::T   # boundary between quadratic and linear loss
+    function HuberLoss(d::T)
+        d > 0 || error("Huber crossover parameter must be strictly positive.")
+        new(d)
+    end
+end
+HuberLoss{T<:AbstractFloat}(d::T=1.0) = HuberLoss{T}(d)
+HuberLoss{T<:Number}(d::T) = HuberLoss{Float64}(Float64(d))
+
+function value{T1,T2<:Number}(loss::HuberLoss{T1}, difference::T2)
+    T = promote_type(T1,T2)
+    abs_diff = abs(difference)
+    if abs_diff < loss.d
+        return T(0.5)*abs2(difference)   # quadratic
+    else
+        return (loss.d*abs_diff) - T(0.5)*abs2(loss.d)   # linear
+    end
+end
+function deriv{T1,T2<:Number}(loss::HuberLoss{T1}, difference::T2)
+    T = promote_type(T1,T2)
+    if abs(difference) < loss.d
+        return T(difference)   # quadratic
+    else
+        return loss.d*T(sign(difference))   # linear
+    end
+end
+function deriv2{T1,T2<:Number}(loss::HuberLoss{T1}, difference::T2)
+    T = promote_type(T1,T2)
+    if abs(difference) < loss.d
+        return one(T)   # quadratic
+    else
+        return zero(T)   # linear
+    end
+end
+function value_deriv{T1,T2<:Number}(loss::HuberLoss{T1}, difference::T2)
+    T = promote_type(T1,T2)
+    abs_diff = abs(difference)
+    if abs_diff < loss.d
+        val = T(0.5)*abs2(difference)
+        der = T(difference)
+    else
+        val = (loss.d*abs_diff) - T(0.5)*abs2(loss.d)
+        der = loss.d*T(sign(difference))
+    end
+    return val,der
+end
+
+isdifferentiable(::HuberLoss) = false
+isdifferentiable(l::HuberLoss, at) = at != abs(l.d)
```

Comment:
Isn't differentiability of the Huber loss function at the origin one of its benefits? :)